### PR TITLE
Add TypeScript definitions

### PR DESCRIPTION
**Pull request by hjkcai**
This is a WIP PR but I want to post it before it completes because **I am not sure my definitions 100% match the API**. 

Every definition comes from the English version in the official documentation [http://element.eleme.io/](http://element.eleme.io/). But the documentation is too ambiguous to get a detailed definition. So I think more discussions are needed!

Please also help me to solve the 'todos' (lines start with `// TODO`). Those todos are the problems I have encountered so far.

TODOs left:

- [x] `date-picker.d.ts` Description of `DateRange`
- [x] `select.d.ts` Use a better interface name of `AlterItemsEventHandler`
- [ ] `table-column.d.ts` Complete type `TableColumn` and `ElTableColumn.filteredValue`
- [ ] `tree.d.ts` Complete type and description of `FilterNodeMethod`
- [ ] `upload.d.ts` Full review needed

### CODE REVIEW

**Comment by hjkcai** [comment]
Related issues: #2651 #2849

File: types/loading.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    fullscreen: boolean
+  }
+}
+
+/** Show animation while loading data */
+export interface ElLoading {
+  /** Install Loading directive into Vue */
+  install (vue: typeof Vue)
+
+  // TODO: add description
```
Comment (Leopoldthecoder) [inline types/loading.d.ts]: If you do not have a specific DOM node to attach the Loading directive, or if you simply prefer not to use Loading as a directive, you can call this service with some configs to open a Loading instance.

**Comment by Leopoldthecoder** [comment]
Thank you so much for this PR. Don't hesitate to ask if you find anything in the docs unclear.

**Comment by hjkcai** [comment]
抱歉这个月很忙…没啥时间来打理这个PR。现在主要问题还是那几个Todo没有解决，特别是Table和Tree的一些内部类型的类型定义。这个应该只有Table和Tree的作者能比较清楚地知道，所以希望大家能一起帮忙完成这个PR。

File: types/select.d.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import { ElementUIComponent, ElementUIComponentSize } from './component'
+
+// TODO: update the name of this interface
+export interface AlterItemsEventHandler {
```
Comment (Leopoldthecoder) [inline types/select.d.ts]: How about `QueryChangeHandler`?

File: types/date-picker.d.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  /**
+   * Determine if `date` will be disabled in the picker
+   *
+   * @param date The date to check
+   * @returns if `date` will be disabled in the picker
+   */
+  (date: Date): boolean
+}
+
+// TODO: add description
```
Comment (Leopoldthecoder) [inline types/date-picker.d.ts]: Picked date range

**Comment by peterchealse** [comment]
什么时候合到element...

**Comment by develar** [comment]
Thanks a lot for typings.

It will be cool to merge step by step and do not wait when all will be ready. For now I just copy files from this PR and concatenated to one d.ts file.

**Comment by spt110** [comment]
什么时候合并

**Comment by rileyio** [comment]
@hjkcai would you be willing to submit your typings to https://github.com/DefinitelyTyped/DefinitelyTyped while we wait for the PR to be merged here?

**Comment by Hanruis** [comment]
I can understand that component(plugin) like `dialog`,`message` can work perfect with these d.ts.
but how the component like 'button', 'table'  benefit from the d.ts .
These components  just be used in html rather than in JavaScript.

**Comment by doxiaodong** [comment]
不加入 test 么

**Comment by Leopoldthecoder** [comment]
We'll merge this PR this week.

**Comment by Leopoldthecoder** [comment]
Merging this now, I'll take it from here.

**Comment by Qquanwei** [comment]
Any release?

**Comment by vtroop** [comment]
Hi, @hjkcai , i update element-ui@2.0.2 and vue@2.5+ and typescript@2.5.3

However, it can't work well with this d.ts
```javascript
 import  ElementUI from 'element-ui'

Vue.use(ElementUI)
```

it will compile without no error, but in the chrome console, it will alert 'can't find property "install" of undefined'

**Comment by hjkcai** [comment]
@vtroop Since there is no default export in element-ui, you need to import all from element-ui like this:

```javascript
import * as ElementUI from 'element-ui'
Vue.use(ElementUI)
```

**Comment by vtroop** [comment]
@hjkcai  However, i have tried on your solution before
it didn't work for me, it will alert another error
```
error TS2345: Argument of type 'typeof "C:/Users/vtroo/Documents/projects/static-api-conf/node_modules/element-ui/types/index"' is not assignable to param
eter of type 'PluginObject<any> | PluginFunction<any>'.
  Type 'typeof "C:/Users/vtroo/Documents/projects/static-api-conf/node_modules/element-ui/types/index"' is not assignable to type 'PluginFunction<any>'.
    Type 'typeof "C:/Users/vtroo/Documents/projects/static-api-conf/node_modules/element-ui/types/index"' provides no match for the signature '(Vue: VueConstructo
r<Vue>, options?: any): void'.
```

**Comment by hjkcai** [comment]
@vtroop Oops.. Seems there is something wrong with these d.ts files.

Here is a quick fix: Open `node_modules/element-ui/types/index.d.ts` and modify the last line.

The original content of that line is:

```typescript
export default ElementUI
```

You need to change it to:

```typescript
export = ElementUI
```

And then wait for the next release ;)

**Comment by Leopoldthecoder** [comment]
@hjkcai Hi, I'm not familiar with TypeScript, so should I change `export default ElementUI` to `export = ElementUI`? I thought if we `export default ElementUI`, we can import it like this:
```JS
import ElementUI from 'element-ui'
```

Am I wrong?

**Comment by hjkcai** [comment]
@Leopoldthecoder See #7939 and #7940

**Comment by vtroop** [comment]
@hjkcai Thx, it works fine for me
```javascript
main.ts
import * as element from 'element-ui'
Vue.use(element)

t.ds
export = ElementUI
```

**Comment by yshskl** [comment]
请问，我使用vue-cli4.0手动创建了一个项目，使用了typescript功能，又引入了elementUI插件（vue-cli-plugin-element），此时项目文件夹内变化如下：

src/plugins/element.js:
```
import Vue from 'vue'
import Element from 'element-ui'
import '../element-variables.scss'

Vue.use(Element)
```

src/main.ts:
```
...
import './plugins/element.js'
...
```

还需要做哪些配置才可以使用elementUI的types类型提示功能？
譬如我使用vscode，开发后台管理系统，使用常用的表格table组件，给它的cell-style属性绑定自定义方法时，鼠标放到方法上就能显示该方法的类型定义之类的；或者像使用form组件的validate方法，vscode能自动识别ref出来的方法，不会报错，或者不需要再手动到types文件夹里找对应方法的types引进来使用。这种设想能实现吗，还是elementUI的类型提示是在别的环节才管用的？

**Pull request merged**